system_zero/src/env/mock.rs

```diff
@@ -0,0 +1,82 @@
+use std::collections::HashMap;
+
+use primitive_types::{H160, H256, U256};
+
+use crate::env::environment::Environment;
+
+pub struct MockEnvironment {
+    accounts: HashMap<H160, MockAccount>,
+}
+
+impl MockEnvironment {
+    pub fn new() -> Self {
+        Self {
+            accounts: HashMap::new(),
+        }
+    }
+}
+
+impl Environment for MockEnvironment {
+    fn get_balance(&self, addr: H160) -> U256 {
+        self.accounts
+            .get(&addr)
+            .map(|acc| acc.balance)
+            .unwrap_or(U256::zero())
+    }
+
+    fn add_balance(&mut self, addr: H160, value: U256) {
+        let acc = self.accounts.entry(addr).or_default();
+        acc.balance += value;
+    }
+
+    fn get_code(&self, addr: H160) -> Vec<u64> {
+        self.accounts
+            .get(&addr)
+            .map(|acc| acc.code.clone())
+            .unwrap_or_else(Vec::new)
+    }
+
+    fn get_code_size(&self, addr: H160) -> usize {
+        self.get_code(addr).len()
+    }
+
+    fn get_code_range(&self, addr: H160, offset: usize, len: usize) -> Vec<u64> {
+        let code = self.get_code(addr);
+        if code.len() < offset {
+            return vec![];
+        }
+        code[offset..(offset + len).min(code.len())].to_vec()
+    }
+
+    fn read_storage(&self, addr: H160, key: H256) -> H256 {
+        let acc = self.accounts.get(&addr).expect("No such address");
+        acc.storage.get(&key).copied().unwrap_or(H256::zero())
+    }
+
+    fn write_storage(&mut self, addr: H160, key: H256, value: H256) {
+        let acc = self.accounts.get_mut(&addr).expect("No such address");
+        acc.storage.insert(key, value);
+    }
+
+    fn create(&mut self, addr: H160, endowment: U256, code: Vec<u64>) {
+        let acc = self.accounts.entry(addr).or_default();
+        acc.code = code;
+        acc.balance += endowment;
+    }
+}
+
+struct MockAccount {
+    balance: U256,
+    code: Vec<u64>,
+    storage: HashMap<H256, H256>,
+}
```

Comment:
Hm, from the yellowpaper, I sort of got the impression that it might be best to view them as one unified model. I.e. I believe it's possible to send ETH to an arbitrary address, in which case we'd have a non-empty account which could be either a EOA of a contract, but we won't know which until someone deploys code there or signs a txn with that address as "from".

I think there's also a sort of "in between" state when a contract's initial code (typically a constructor) is being run, so then we know it is (will be?) a contract, but it's code is empty until the initial code returns (and could remain empty if it doesn't return anything).

Reply:
Hm I see, so in this model there is no distinction between
- An EOA with non-zero balance,
- A contract without code and non-zero balance

right? I think that works, although distinguishing between EOA and contract account might require going through the txs history. I wonder if it's easier to make the distinction explicit. Then we could add metadata to the contract accounts, like the creation tx or events.

Reply:
I'm not sure if it's always possible to distinguish them? I.e. consider a value-only transaction where the `to` address doesn't exist yet. In that case `to` is probably a EOA, but technically it could be a contract (I believe it's valid to deploy a contract to an address with non-zero value). I don't think we can tell until some further action is taken.

We could have an enum variant for that ambiguous case, but I think one struct may be easier. FWIW it seems openethereum has one [Account](https://github.com/openethereum/openethereum/blob/main/crates/ethcore/src/state/account.rs#L54) struct.